libs/vkd3d/command.c

```diff
@@ -2692,6 +2696,339 @@ static void d3d12_command_list_handle_active_queries(struct d3d12_command_list *
+             * SSBO alignment issues on some hardware. */
+            resolve_buffer_stride = get_query_heap_stride(q->heap->desc.Type);
+            resolve_buffer_size = align(resolve_buffer_size, ssbo_alignment);
+            resolve_index = 0;
+
+            d = &dispatches[dispatch_count++];
+            d->heap = q->heap;
+            d->virtual_query_count = 1;
+            d->unique_query_count = 0;
+            d->resolve_buffer_offset = resolve_buffer_size;
+
+            r = NULL;
+        }
+        else
+            d->virtual_query_count++;
+
+        /* Prepare one resolve entry per Vulkan query range */
+        if (!r || r->query_pool != q->vk_pool || r->first_query + r->query_count != q->vk_index)
+        {
+            if (!vkd3d_array_reserve((void **)&resolves, &resolve_size, resolve_count + 1, sizeof(*resolves)))
+            {
+                ERR("Failed to allocate resolve list.\n");
+                goto cleanup;
+            }
+
+            r = &resolves[resolve_count++];
+            r->query_pool = q->vk_pool;
+            r->first_query = q->vk_index;
+            r->query_count = 1;
+            r->offset = resolve_buffer_size;
+            r->stride = get_query_heap_stride(q->heap->desc.Type);
+        }
+        else
+            r->query_count++;
+
+        query_map_size = max(query_map_size, q->index + 1);
+        resolve_buffer_size += resolve_buffer_stride;
+
+        d->resolve_buffer_size = resolve_buffer_size - d->resolve_buffer_offset;
+        q->resolve_index = resolve_index++;
+    }
+
+    /* Allocate scratch buffer and resolve virtual Vulkan queries into it */
+    if (!d3d12_command_allocator_allocate_scratch_memory(list->allocator,
+            resolve_buffer_size, ssbo_alignment, &resolve_buffer))
+        goto cleanup;
+
+    for (i = 0; i < resolve_count; i++)
+    {
+        const struct resolve_entry *r = &resolves[i];
+
+        VK_CALL(vkCmdCopyQueryPoolResults(list->vk_command_buffer,
+            r->query_pool, r->first_query, r->query_count,
+            resolve_buffer.buffer, resolve_buffer.offset + r->offset,
+            r->stride, VK_QUERY_RESULT_WAIT_BIT | VK_QUERY_RESULT_64_BIT));
+    }
+
+    /* Allocate scratch buffer for query lists */
+    entry_buffer_size = sizeof(struct query_entry) * list->pending_queries_count;
+
+    if (!d3d12_command_allocator_allocate_scratch_memory(list->allocator,
+            entry_buffer_size, ssbo_alignment, &entry_buffer))
+        goto cleanup;
+
+    if (!(query_map = vkd3d_malloc(sizeof(*query_map) * query_map_size)) ||
+            !(query_list = vkd3d_malloc(sizeof(*query_list) * list->pending_queries_count)))
+    {
+        ERR("Failed to allocate query map.\n");
+        goto cleanup;
+    }
+
+    /* Active list for the current dispatch */
+    src_queries = list->pending_queries;
+    dst_queries = query_list;
+
+    for (i = 0; i < dispatch_count; i++)
```

Comment:
Quite hard to follow this loop, but I understand the data structure it's trying to emit, so it's probably fine.